# 1. Importing Libraries

In [2]:
import numpy as np
import pandas as pd
import seaborn as sns
import matplotlib.pyplot as plt
import sklearn
import warnings
warnings.filterwarnings('ignore')

# 2. Importing Dataset

In [3]:
df=pd.read_csv('Grades.csv')
df

Seat No. PH-121 HS-101 CY-105 HS-105/12 MT-111 CS-105 CS-106 EL-102  \
0    CS-97001     B-     D+     C-         C     C-     D+      D     C-   
1    CS-97002      A      D     D+         D     B-      C      D      A   
2    CS-97003      A      B      A        B-     B+      A     B-     B+   
3    CS-97004      D     C+     D+         D      D     A-     D+     C-   
4    CS-97005     A-     A-     A-        B+      A      A     A-     B+   
..        ...    ...    ...    ...       ...    ...    ...    ...    ...   
566  CS-97567      B      A      A        A-     A+      A     A-     A-   
567  CS-97568     A+      A      A         A      A      A      A     A-   
568  CS-97569      B      A     A-        B+      A      A      A      A   
569  CS-97570      A     B+      D         A      D     D+     B-     C-   
570  CS-97571      C      D      D         C      C     D+      B     C+   

    EE-119  ... CS-312 CS-317 CS-403 CS-421 CS-406 CS-414 CS-419 CS-423  \
0       B-  ...     C-     C-     C-     C-     A-      A     C-      B   
1       D+  ...     D+      D      C      D     A-     B-      C      C   
2       A-  ...      B      B      A      C      A      A      A     A-   
3        D  ...     D+      C     D+     C-     B-      B     C+     C+   
4        A  ...     B-     B+     B+     B-     A-      A     A-     A-   
..     ...  ...    ...    ...    ...    ...    ...    ...    ...    ...   
566     A+  ...     A-     A-      A      A      A     B+     B+      B   
567      A  ...     B+     B+      A      A     A-      B     A-      C   
568      A  ...     A-      B      A     B+      A      C     B+     A-   
569     B-  ...      D      B      B     C-      D      C      B     B-   
570      C  ...     C+      C     B-      D      F     C-     B+      D   

    CS-412   CGPA  
0       A-  2.205  
1        B  2.008  
2        A  3.608  
3       C+  1.906  
4        A  3.448  
..     ...    ...  
566      A  3.798  
567     A-  3.772  
568     A-  3.470  
569      C  2.193  
570     C-  1.753  

[571 rows x 43 columns]

Since all the columns are alphabetical, we will have to convert the columns to numerical values

# 3. Finding null values

In [4]:
#Before converting the categorical columns to numerical, we need to find the null values and assign any value to NaN
df.isnull().sum()

Seat No.      0
PH-121        0
HS-101        0
CY-105        1
HS-105/12     1
MT-111        2
CS-105        0
CS-106        2
EL-102        2
EE-119        2
ME-107        2
CS-107        2
HS-205/20     5
MT-222        5
EE-222        7
MT-224        7
CS-210        7
CS-211        5
CS-203        5
CS-214        6
EE-217        6
CS-212        6
CS-215        6
MT-331        9
EF-303       10
HS-304       10
CS-301       10
CS-302       10
TC-383       10
MT-442       10
EL-332        9
CS-318        9
CS-306        9
CS-312       10
CS-317       12
CS-403       12
CS-421       12
CS-406       85
CS-414       13
CS-419       13
CS-423       14
CS-412       79
CGPA          0
dtype: int64

In [5]:
for i in df.columns:
    if df[i].isnull().any():
        df[i].fillna(-1,inplace=True)
df

Seat No. PH-121 HS-101 CY-105 HS-105/12 MT-111 CS-105 CS-106 EL-102  \
0    CS-97001     B-     D+     C-         C     C-     D+      D     C-   
1    CS-97002      A      D     D+         D     B-      C      D      A   
2    CS-97003      A      B      A        B-     B+      A     B-     B+   
3    CS-97004      D     C+     D+         D      D     A-     D+     C-   
4    CS-97005     A-     A-     A-        B+      A      A     A-     B+   
..        ...    ...    ...    ...       ...    ...    ...    ...    ...   
566  CS-97567      B      A      A        A-     A+      A     A-     A-   
567  CS-97568     A+      A      A         A      A      A      A     A-   
568  CS-97569      B      A     A-        B+      A      A      A      A   
569  CS-97570      A     B+      D         A      D     D+     B-     C-   
570  CS-97571      C      D      D         C      C     D+      B     C+   

    EE-119  ... CS-312 CS-317 CS-403 CS-421 CS-406 CS-414 CS-419 CS-423  \
0       B-  ...     C-     C-     C-     C-     A-      A     C-      B   
1       D+  ...     D+      D      C      D     A-     B-      C      C   
2       A-  ...      B      B      A      C      A      A      A     A-   
3        D  ...     D+      C     D+     C-     B-      B     C+     C+   
4        A  ...     B-     B+     B+     B-     A-      A     A-     A-   
..     ...  ...    ...    ...    ...    ...    ...    ...    ...    ...   
566     A+  ...     A-     A-      A      A      A     B+     B+      B   
567      A  ...     B+     B+      A      A     A-      B     A-      C   
568      A  ...     A-      B      A     B+      A      C     B+     A-   
569     B-  ...      D      B      B     C-      D      C      B     B-   
570      C  ...     C+      C     B-      D      F     C-     B+      D   

    CS-412   CGPA  
0       A-  2.205  
1        B  2.008  
2        A  3.608  
3       C+  1.906  
4        A  3.448  
..     ...    ...  
566      A  3.798  
567     A-  3.772  
568     A-  3.470  
569      C  2.193  
570     C-  1.753  

[571 rows x 43 columns]

1. We find if NaN values are present in range of columns<br>
2. We fill the NaN values with '-1'.

In [6]:
# To check for null values after filling NaN
df.isnull().sum()

Seat No.     0
PH-121       0
HS-101       0
CY-105       0
HS-105/12    0
MT-111       0
CS-105       0
CS-106       0
EL-102       0
EE-119       0
ME-107       0
CS-107       0
HS-205/20    0
MT-222       0
EE-222       0
MT-224       0
CS-210       0
CS-211       0
CS-203       0
CS-214       0
EE-217       0
CS-212       0
CS-215       0
MT-331       0
EF-303       0
HS-304       0
CS-301       0
CS-302       0
TC-383       0
MT-442       0
EL-332       0
CS-318       0
CS-306       0
CS-312       0
CS-317       0
CS-403       0
CS-421       0
CS-406       0
CS-414       0
CS-419       0
CS-423       0
CS-412       0
CGPA         0
dtype: int64

No null values are present in any column, now we can go ahead and convert the columns using encoding technique

# 3. Encoding the Categorical columns

In [7]:
df.dtypes

Seat No.      object
PH-121        object
HS-101        object
CY-105        object
HS-105/12     object
MT-111        object
CS-105        object
CS-106        object
EL-102        object
EE-119        object
ME-107        object
CS-107        object
HS-205/20     object
MT-222        object
EE-222        object
MT-224        object
CS-210        object
CS-211        object
CS-203        object
CS-214        object
EE-217        object
CS-212        object
CS-215        object
MT-331        object
EF-303        object
HS-304        object
CS-301        object
CS-302        object
TC-383        object
MT-442        object
EL-332        object
CS-318        object
CS-306        object
CS-312        object
CS-317        object
CS-403        object
CS-421        object
CS-406        object
CS-414        object
CS-419        object
CS-423        object
CS-412        object
CGPA         float64
dtype: object

CGPA column is float type<br>
All other columns are object types

In [8]:
from sklearn.preprocessing import OrdinalEncoder
OE=OrdinalEncoder()
for i in df.columns[1:42]:
    df[i]=df[i].astype(str)

for i in df.columns[1:42]:
    df[i]=OE.fit_transform(df[i].values.reshape(-1,1))
df

Seat No.  PH-121  HS-101  CY-105  HS-105/12  MT-111  CS-105  CS-106  \
0    CS-97001     5.0    10.0     9.0        7.0     9.0    10.0    10.0   
1    CS-97002     0.0     9.0    11.0       10.0     6.0     6.0    10.0   
2    CS-97003     0.0     3.0     1.0        6.0     5.0     0.0     6.0   
3    CS-97004     9.0     7.0    11.0       10.0    10.0     2.0    11.0   
4    CS-97005     2.0     2.0     3.0        5.0     1.0     0.0     3.0   
..        ...     ...     ...     ...        ...     ...     ...     ...   
566  CS-97567     3.0     0.0     1.0        3.0     2.0     0.0     3.0   
567  CS-97568     1.0     0.0     1.0        1.0     1.0     0.0     1.0   
568  CS-97569     3.0     0.0     3.0        5.0     1.0     0.0     1.0   
569  CS-97570     0.0     4.0    10.0        1.0    10.0    10.0     6.0   
570  CS-97571     6.0     9.0    10.0        7.0     7.0    10.0     4.0   

     EL-102  EE-119  ...  CS-312  CS-317  CS-403  CS-421  CS-406  CS-414  \
0       9.0     6.0  ...     9.0     9.0     9.0     9.0     3.0     1.0   
1       1.0    11.0  ...    11.0    10.0     7.0    10.0     3.0     6.0   
2       5.0     3.0  ...     4.0     4.0     1.0     7.0     1.0     1.0   
3       9.0    10.0  ...    11.0     7.0    11.0     9.0     6.0     4.0   
4       5.0     1.0  ...     6.0     5.0     5.0     6.0     3.0     1.0   
..      ...     ...  ...     ...     ...     ...     ...     ...     ...   
566     3.0     2.0  ...     3.0     3.0     1.0     1.0     1.0     5.0   
567     3.0     1.0  ...     5.0     5.0     1.0     1.0     3.0     4.0   
568     1.0     1.0  ...     3.0     4.0     1.0     5.0     1.0     7.0   
569     9.0     6.0  ...    10.0     4.0     4.0     9.0    10.0     7.0   
570     8.0     7.0  ...     8.0     7.0     6.0    10.0    12.0     9.0   

     CS-419  CS-423  CS-412   CGPA  
0       9.0     4.0     3.0  2.205  
1       7.0     7.0     4.0  2.008  
2       1.0     3.0     1.0  3.608  
3       8.0     8.0     8.0  1.906  
4       3.0     3.0     1.0  3.448  
..      ...     ...     ...    ...  
566     5.0     4.0     1.0  3.798  
567     3.0     7.0     3.0  3.772  
568     5.0     3.0     3.0  3.470  
569     4.0     6.0     7.0  2.193  
570     5.0    10.0     9.0  1.753  

[571 rows x 43 columns]

Column 'Seat No' has all the unique values and hence can be dropped

# 4. Exploratory Data Analysis

In [9]:
#1. To find the top 5 rows in the dataset
df.head(5)

Seat No.  PH-121  HS-101  CY-105  HS-105/12  MT-111  CS-105  CS-106  \
0  CS-97001     5.0    10.0     9.0        7.0     9.0    10.0    10.0   
1  CS-97002     0.0     9.0    11.0       10.0     6.0     6.0    10.0   
2  CS-97003     0.0     3.0     1.0        6.0     5.0     0.0     6.0   
3  CS-97004     9.0     7.0    11.0       10.0    10.0     2.0    11.0   
4  CS-97005     2.0     2.0     3.0        5.0     1.0     0.0     3.0   

   EL-102  EE-119  ...  CS-312  CS-317  CS-403  CS-421  CS-406  CS-414  \
0     9.0     6.0  ...     9.0     9.0     9.0     9.0     3.0     1.0   
1     1.0    11.0  ...    11.0    10.0     7.0    10.0     3.0     6.0   
2     5.0     3.0  ...     4.0     4.0     1.0     7.0     1.0     1.0   
3     9.0    10.0  ...    11.0     7.0    11.0     9.0     6.0     4.0   
4     5.0     1.0  ...     6.0     5.0     5.0     6.0     3.0     1.0   

   CS-419  CS-423  CS-412   CGPA  
0     9.0     4.0     3.0  2.205  
1     7.0     7.0     4.0  2.008  
2     1.0     3.0     1.0  3.608  
3     8.0     8.0     8.0  1.906  
4     3.0     3.0     1.0  3.448  

[5 rows x 43 columns]

All the columns(apart from CGPA) after transformation are float types<br>
CGPA column is continous data and there are not many unique numbers present<br>
It is a problem of Linear Regression

In [10]:
#2. To find the shape of the dataset
df.shape

(571, 43)

Rows : 571<br>
Columns : 43<br>

In [11]:
#3. To find the column names
df.columns

Index(['Seat No.', 'PH-121', 'HS-101', 'CY-105', 'HS-105/12', 'MT-111',
       'CS-105', 'CS-106', 'EL-102', 'EE-119', 'ME-107', 'CS-107', 'HS-205/20',
       'MT-222', 'EE-222', 'MT-224', 'CS-210', 'CS-211', 'CS-203', 'CS-214',
       'EE-217', 'CS-212', 'CS-215', 'MT-331', 'EF-303', 'HS-304', 'CS-301',
       'CS-302', 'TC-383', 'MT-442', 'EL-332', 'CS-318', 'CS-306', 'CS-312',
       'CS-317', 'CS-403', 'CS-421', 'CS-406', 'CS-414', 'CS-419', 'CS-423',
       'CS-412', 'CGPA'],
      dtype='object')

In [12]:
#4. To find the column types
df.dtypes

Seat No.      object
PH-121       float64
HS-101       float64
CY-105       float64
HS-105/12    float64
MT-111       float64
CS-105       float64
CS-106       float64
EL-102       float64
EE-119       float64
ME-107       float64
CS-107       float64
HS-205/20    float64
MT-222       float64
EE-222       float64
MT-224       float64
CS-210       float64
CS-211       float64
CS-203       float64
CS-214       float64
EE-217       float64
CS-212       float64
CS-215       float64
MT-331       float64
EF-303       float64
HS-304       float64
CS-301       float64
CS-302       float64
TC-383       float64
MT-442       float64
EL-332       float64
CS-318       float64
CS-306       float64
CS-312       float64
CS-317       float64
CS-403       float64
CS-421       float64
CS-406       float64
CS-414       float64
CS-419       float64
CS-423       float64
CS-412       float64
CGPA         float64
dtype: object

All the columns are numericals columns now

In [13]:
df.nunique().to_frame('No.of unique values')

No.of unique values
Seat No.                   571
PH-121                      13
HS-101                      12
CY-105                      14
HS-105/12                   14
MT-111                      14
CS-105                      11
CS-106                      14
EL-102                      14
EE-119                      13
ME-107                      14
CS-107                      15
HS-205/20                   14
MT-222                      15
EE-222                      14
MT-224                      15
CS-210                      15
CS-211                      15
CS-203                      14
CS-214                      14
EE-217                      14
CS-212                      13
CS-215                      15
MT-331                      13
EF-303                      13
HS-304                      15
CS-301                      13
CS-302                      12
TC-383                      13
MT-442                      13
EL-332                      14
CS-318                      15
CS-306                      14
CS-312                      15
CS-317                      13
CS-403                      12
CS-421                      14
CS-406                      15
CS-414                      14
CS-419                      13
CS-423                      13
CS-412                      14
CGPA                       491

Column "Seat No." has all the rows as unique values and can be dropped, since it doesn't contribute to the CGPA.

In [14]:
df.drop('Seat No.', axis=1, inplace=True)

In [15]:
#Finding number of unique columns after dropping 1st Column
df.nunique().to_frame('No.of unique values')

No.of unique values
PH-121                      13
HS-101                      12
CY-105                      14
HS-105/12                   14
MT-111                      14
CS-105                      11
CS-106                      14
EL-102                      14
EE-119                      13
ME-107                      14
CS-107                      15
HS-205/20                   14
MT-222                      15
EE-222                      14
MT-224                      15
CS-210                      15
CS-211                      15
CS-203                      14
CS-214                      14
EE-217                      14
CS-212                      13
CS-215                      15
MT-331                      13
EF-303                      13
HS-304                      15
CS-301                      13
CS-302                      12
TC-383                      13
MT-442                      13
EL-332                      14
CS-318                      15
CS-306                      14
CS-312                      15
CS-317                      13
CS-403                      12
CS-421                      14
CS-406                      15
CS-414                      14
CS-419                      13
CS-423                      13
CS-412                      14
CGPA                       491

In [16]:
for i in df.columns:
    print(df[i].value_counts())
    print('\n')

2.0     112
0.0     111
4.0      61
3.0      57
5.0      56
9.0      44
6.0      33
7.0      31
10.0     22
1.0      22
8.0      19
12.0      2
11.0      1
Name: PH-121, dtype: int64


2.0     82
5.0     78
6.0     68
3.0     63
4.0     59
8.0     50
7.0     47
9.0     45
0.0     38
10.0    36
1.0      4
11.0     1
Name: HS-101, dtype: int64


1.0     177
3.0     120
5.0      50
4.0      49
6.0      42
10.0     31
2.0      31
7.0      19
8.0      17
9.0      16
11.0     14
13.0      3
0.0       1
12.0      1
Name: CY-105, dtype: int64


1.0     96
3.0     75
5.0     70
4.0     57
10.0    45
7.0     41
6.0     40
8.0     39
9.0     36
11.0    34
2.0     34
13.0     2
0.0      1
12.0     1
Name: HS-105/12, dtype: int64


3.0     105
1.0     100
6.0      70
5.0      62
4.0      55
9.0      39
8.0      33
7.0      30
10.0     26
2.0      23
11.0     21
13.0      3
0.0       2
12.0      2
Name: MT-111, dtype: int64


0.0     151
2.0     134
4.0      60
3.0      51
1.0      43
5.0      38
7.

There is no white spaces or blank spaces in any column<br>
In column CGPA, out of 571 rows, 491 are unique values, this shows at almost all the value as unique value and no category is present

In [17]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 571 entries, 0 to 570
Data columns (total 42 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   PH-121     571 non-null    float64
 1   HS-101     571 non-null    float64
 2   CY-105     571 non-null    float64
 3   HS-105/12  571 non-null    float64
 4   MT-111     571 non-null    float64
 5   CS-105     571 non-null    float64
 6   CS-106     571 non-null    float64
 7   EL-102     571 non-null    float64
 8   EE-119     571 non-null    float64
 9   ME-107     571 non-null    float64
 10  CS-107     571 non-null    float64
 11  HS-205/20  571 non-null    float64
 12  MT-222     571 non-null    float64
 13  EE-222     571 non-null    float64
 14  MT-224     571 non-null    float64
 15  CS-210     571 non-null    float64
 16  CS-211     571 non-null    float64
 17  CS-203     571 non-null    float64
 18  CS-214     571 non-null    float64
 19  EE-217     571 non-null    float64
 20  CS-212    

All the columns are non-null columns<br>
Total Columns : 42<br>
Numerical columns : 41 <br>
Object columns : 1 <br>

# 5. Description of Dataset

In [18]:
df.describe()

PH-121      HS-101      CY-105   HS-105/12      MT-111      CS-105  \
count  571.000000  571.000000  571.000000  571.000000  571.000000  571.000000   
mean     3.781086    5.071804    3.896673    5.239930    4.886165    2.838879   
std      3.046895    2.785317    2.966747    3.203104    3.000468    2.696709   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%      2.000000    3.000000    1.000000    3.000000    3.000000    0.000000   
50%      3.000000    5.000000    3.000000    5.000000    5.000000    2.000000   
75%      6.000000    7.000000    5.500000    8.000000    7.000000    4.000000   
max     12.000000   11.000000   13.000000   13.000000   13.000000   10.000000   

           CS-106      EL-102      EE-119      ME-107  ...      CS-312  \
count  571.000000  571.000000  571.000000  571.000000  ...  571.000000   
mean     5.112084    4.949212    4.875657    5.768827  ...    5.036778   
std      2.741119    3.043407    2.670886    3.160470  ...    3.404390   
min      0.000000    0.000000    0.000000    0.000000  ...    0.000000   
25%      3.000000    3.000000    3.000000    3.000000  ...    2.000000   
50%      5.000000    4.000000    4.000000    6.000000  ...    4.000000   
75%      6.000000    7.000000    7.000000    9.000000  ...    8.000000   
max     13.000000   13.000000   12.000000   13.000000  ...   14.000000   

           CS-317      CS-403      CS-421      CS-406      CS-414      CS-419  \
count  571.000000  571.000000  571.000000  571.000000  571.000000  571.000000   
mean     5.709282    4.411559    6.301226    3.537653    3.374781    4.788091   
std      2.865322    3.008259    2.859944    2.848804    2.525587    2.476330   
min      0.000000    0.000000    0.000000    0.000000    0.000000    0.000000   
25%      4.000000    1.000000    4.000000    1.000000    1.000000    3.000000   
50%      6.000000    4.000000    6.000000    3.000000    3.000000    5.000000   
75%      8.000000    6.000000    8.000000    5.000000    5.000000    6.000000   
max     12.000000   11.000000   13.000000   14.000000   13.000000   12.000000   

           CS-423      CS-412        CGPA  
count  571.000000  571.000000  571.000000  
mean     4.893170    3.633975    2.954888  
std      2.884466    2.717875    0.620552  
min      0.000000    0.000000    0.800000  
25%      3.000000    1.000000    2.538000  
50%      5.000000    3.000000    3.029000  
75%      7.000000    5.000000    3.451000  
max     12.000000   13.000000    3.985000  

[8 rows x 42 columns]

Statistical information of numerical columns:<br>
Mean>Median : <br>
Mean<Median : <br>
Difference between 3rd quartile and Maximum value : <br>